daemon/src/payout_curve.rs

```diff
@@ -1,30 +1,794 @@
-use anyhow::Result;
+use crate::payout_curve::curve::Curve;
+
+use anyhow::{Context, Result};
 use bdk::bitcoin;
-use cfd_protocol::interval::MAX_PRICE_DEC;
 use cfd_protocol::{generate_payouts, Payout};
+use ndarray::prelude::*;
+use num::{FromPrimitive, ToPrimitive};
+use rust_decimal::Decimal;
+
+mod basis;
+mod basis_eval;
+mod compat;
+mod csr_tools;
+mod curve;
+mod curve_factory;
+mod splineobject;
+mod utils;
 
+pub enum LongPosition {
+    #[allow(dead_code)]
+    Maker,
+    Taker,
+}
+
+/// function to generate an iterator of values, heuristically viewed as:
+///
+///     `[left_price_boundary, right_price_boundary], maker_payout_value`
+///
+/// with units
+///
+///     `[Usd, Usd], bitcoin::Amount`
+///
+/// A key item to note is that although the POC logic has been to imposed
+/// that maker goes short every time, there is no reason to make the math
+/// have this imposition as well. As such, the `long_position` parameter
+/// is used to indicate which party (Maker or Taker) has the long position,
+/// and everything else is handled internally.
+///
+/// As well, the POC has also demanded that the Maker always has unity
+/// leverage, hence why the ability to to specify this amount has been
+/// omitted from the parameters. Internally, it is hard-coded to unity
+/// in the call to PayoutCurve::new(), so this behaviour can be changed in
+/// the future trivially.
+///
+/// ### paramters
+/// * price: BTC-USD exchange rate used to create CFD contract
+/// * quantity: Interger number of one-dollar USD contracts contained in the
+/// CFD; expressed as a Usd amount
+/// * long_position: Indicates which party (Maker or Taker) has taken the long
+/// position for the CFD
+/// * taker_leverage: Leveraging used by the taker
+///
+/// ### returns
+/// * tuple: first element is the iterator described above, second element is
+/// the total amount of BTC that can change ownership via this model. That is,
+/// `taker_payout = total - maker_payout`
 pub fn calculate(
     price: Usd,
-    _quantity: Usd,
-    maker_payin: bitcoin::Amount,
-    (taker_payin, _leverage): (bitcoin::Amount, Leverage),
-) -> Result<Vec<Payout>> {
-    let dollars = price.try_into_u64()?;
-    let payouts = vec![
-        generate_payouts(
-            0..=(dollars - 10),
-            maker_payin + taker_payin,
-            bitcoin::Amount::ZERO,
+    quantity: Usd,
+    long_position: LongPosition,
+    taker_leverage: Leverage,
+) -> Result<(Vec<Payout>, bitcoin::Amount)> {
+    let contract_value = 1_f64;
+    let initial_rate = price
+        .try_into_u64()
+        .context("Cannot convert price to u64")? as f64;
+    let quantity = quantity
+        .try_into_u64()
+        .context("Cannot convert quantity to u64")? as usize;
+    let maker_leverage = 1;
+
+    let maker_payout_curve = match long_position {
+        LongPosition::Maker => PayoutCurve::new(
+            initial_rate as f64,
+            maker_leverage,
+            taker_leverage.0 as usize,
```

Comment:
This seems to be switched: 

Second argument says: `leverage_long` but `maker_leverage` is provided who is short. and third argument says `leverage_short` but `taker_leverage` is provided who is long.

Or am I misreading something?